src/backend/distributed/connection/connection_management.c

```diff
@@ -445,15 +603,102 @@ ShutdownConnection(MultiConnection *connection)
 void
 FinishConnectionListEstablishment(List *multiConnectionList)
 {
+	int connectionIndex = 0;
+	int connectionCount = 0;
+	MultiConnectionReady *connections = NULL;
 	ListCell *multiConnectionCell = NULL;
 
+	WaitEventSet *waitEventSet = NULL;
+	bool waitEventSetRebuild = false;
+	int waitCount = 0;
+	WaitEvent *events = NULL;
+
+	/*
+	 * Here we prepare an array that keeps the MultiConnection together with the
+	 * information on if and how to wait before PQconnectionPoll should be called on it
+	 */
+	connectionCount = list_length(multiConnectionList);
+	connections = (MultiConnectionReady *) palloc0(connectionCount *
+												   sizeof(MultiConnectionReady));
 	foreach(multiConnectionCell, multiConnectionList)
 	{
-		MultiConnection *multiConnection = (MultiConnection *) lfirst(
-			multiConnectionCell);
+		MultiConnection *connection = (MultiConnection *) lfirst(multiConnectionCell);
+
+		connections[connectionIndex].connection = connection;
+
+		/*
+		 * before we can build the waitset to wait for asynchronous IO we need to know the
+		 * pollmode to use for the sockets. This is populated by executing one round of
+		 * PQconnectPoll. This updates the MultiConnectionReady struct with the ready state
+		 * and its next poll mode.
+		 */
+		MultiConnectionReadyExecutePoll(&connections[connectionIndex]);
+
+		if (!connections[connectionIndex].ready)
+		{
+			waitCount++;
+		}
+
+		connectionIndex++;
+	}
 
-		/* TODO: consider making connection establishment fully in parallel */
-		FinishConnectionEstablishment(multiConnection);
+	/* prepapre space for socket events */
+	events = (WaitEvent *) palloc0(connectionCount * sizeof(WaitEvent));
+
+	while (waitCount > 0)
+	{
+		long timeout = -1;
+		int eventCount = 0;
+		int eventIndex = 0;
+
+		if (waitEventSet == NULL || waitEventSetRebuild)
+		{
+			waitEventSet = WaitEventSetFromMultiConnectionReadyArray(connections,
+																	 connectionCount,
+																	 &waitCount);
+
+			/* TODO use resource that only holds the last pointer and frees it on swap */
+			EnsureReleaseResource((MemoryContextCallbackFunction) (&FreeWaitEventSet),
+								  waitEventSet);
+
+			if (waitCount <= 0)
+			{
+				break;
+			}
+		}
+
+		eventCount = WaitEventSetWait(waitEventSet, timeout, events, waitCount,
+									  WAIT_EVENT_CLIENT_READ);
+
```

Comment:
should probably check for a 0 return value and then check the timeout

Reply:
Makes sense, I think we should also set the `timeout` in that case. Currently its -1, but I think we could calculate a deadline based on the citus connection timeout setting and compute the timeout based on the deadline.